### ISSUE TITLE
DynamoDB table not replaced when changing name or range_key of local_secondary_index

### ISSUE THREAD

**Issue by tjoris** [terraform-providers/terraform-provider-aws#12334]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

* Terraform v0.12.18
* provider.aws v2.52.0

### Affected Resource(s)

* aws_dynamodb_table

### Terraform Configuration Files

```hcl
resource "aws_dynamodb_table" "tasks_dynamodb_table_promotionblobcleanup" {
  name         = "promotionblobcleanup"
  billing_mode = "PAY_PER_REQUEST"

  hash_key  = "keyspace"
  range_key = "blob_id"

  attribute {
    name = "keyspace"
    type = "S"
  }

  attribute {
    name = "blob_id"
    type = "S"
  }

  attribute {
    name = "created"
    type = "S"
  }

  local_secondary_index {
    name            = "lsi-created"
    range_key       = "created"
    projection_type = "ALL"
  }

  tags = {
    Project     = var.project
    Environment = var.environment
  }
}
```

### Expected Behavior

The DynamoDB table should be replaced.

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/LSI.html

> Local secondary indexes on a table are created when the table is created.

https://www.terraform.io/docs/providers/aws/r/dynamodb_table.html#local_secondary_index

> these can only be allocated at creation so you cannot change this definition after you have created the resource

Since a local secondary index can't be changed after a table is created, the table should be recreated when there is a change in one of the local secondary indexes.

This is the expected terraform output:

```hcl
  # aws_dynamodb_table.tasks_dynamodb_table_promotionblobcleanup must be replaced
-/+ resource "aws_dynamodb_table" "tasks_dynamodb_table_promotionblobcleanup" {
      ~ arn              = "arn:aws:dynamodb:eu-west-1:067346078311:table/promotionblobcleanup" -> (known after apply)
        billing_mode     = "PAY_PER_REQUEST"
        hash_key         = "keyspace"
      ~ id               = "promotionblobcleanup" -> (known after apply)
        name             = "promotionblobcleanup"
        range_key        = "blob_id"
      - read_capacity    = 0 -> null
      + stream_arn       = (known after apply)
      - stream_enabled   = false -> null
      + stream_label     = (known after apply)
      + stream_view_type = (known after apply)
        tags             = {
            "Environment" = "development"
            "Project"     = "secp"
        }
      - write_capacity   = 0 -> null

        attribute {
            name = "blob_id"
            type = "S"
        }
        attribute {
            name = "created"
            type = "S"
        }
        attribute {
            name = "keyspace"
            type = "S"
        }

      - local_secondary_index { # forces replacement
          - name               = "lsi-created" -> null
          - non_key_attributes = [] -> null
          - projection_type    = "ALL" -> null
          - range_key          = "created" -> null
        }
      + local_secondary_index { # forces replacement
          + name               = "si-created"
          + non_key_attributes = []
          + projection_type    = "ALL"
          + range_key          = "created"
        }

      ~ point_in_time_recovery {
          ~ enabled = false -> (known after apply)
        }

      + server_side_encryption {
          + enabled     = (known after apply)
          + kms_key_arn = (known after apply)
        }

      - ttl {
          - enabled = false -> null
        }
    }
```

### Actual Behavior

The DynamoDB table is updated in-place.

But the local secondary index's name remains the same after the `terraform apply`, because you can't make changes to a local secondary index after the table was created.

This is the actual terraform output:

```hcl
  # aws_dynamodb_table.tasks_dynamodb_table_promotionblobcleanup will be updated in-place
  ~ resource "aws_dynamodb_table" "tasks_dynamodb_table_promotionblobcleanup" {
        arn            = "arn:aws:dynamodb:eu-west-1:067346078311:table/promotionblobcleanup"
        billing_mode   = "PAY_PER_REQUEST"
        hash_key       = "keyspace"
        id             = "promotionblobcleanup"
        name           = "promotionblobcleanup"
        range_key      = "blob_id"
        read_capacity  = 0
        stream_enabled = false
        tags           = {
            "Environment" = "development"
            "Project"     = "secp"
        }
        write_capacity = 0

        attribute {
            name = "blob_id"
            type = "S"
        }
        attribute {
            name = "created"
            type = "S"
        }
        attribute {
            name = "keyspace"
            type = "S"
        }

      - local_secondary_index {
          - name               = "lsi-created" -> null
          - non_key_attributes = [] -> null
          - projection_type    = "ALL" -> null
          - range_key          = "created" -> null
        }
      + local_secondary_index {
          + name               = "si-created"
          + non_key_attributes = []
          + projection_type    = "ALL"
          + range_key          = "created"
        }

        point_in_time_recovery {
            enabled = false
        }

        ttl {
            enabled = false
        }
    }

```

### Steps to Reproduce

1. `terraform apply`
1. in the terraform configuration file, change the name of the local secondary index from "lsi-created" to "si-created"
1. `terraform apply`

**Comment by tjoris** [comment]
I tried to add a test in
https://github.com/terraform-providers/terraform-provider-aws/pull/12335/commits/e3d8a7c48444cc98578dcaff18e080e56aa83873
Is there something else that I can do to help progress this issue?

**Comment by jeffbski-rga** [comment]
I am also seeing this problem when I try to update an LSI index. However I found as a work around that I can add a new LSI index and it will pick up the change. I just cannot update an existing index. So that is unfortunate since now I have duplicate indexes and extra work for dynamo to update them, but I can continue until this is resolved.

**Comment by bflad** [comment]
@anGie44 while you're in the LSI neighborhood, I believe this one is pretty straightforward where `ForceNew` is not present on all the configuration block nested attributes which the associated PR #12335 is fixing 👍

**Comment by anGie44** [comment]
The PR fixing this behavior has been merged and will release with `v3.8.0` of the Terraform AWS Provider, likely out next Thursday.
